barretenberg/cpp/src/barretenberg/vm/avm_trace/avm_execution.cpp

```diff
@@ -35,7 +37,37 @@ HonkProof Execution::run_and_prove(std::vector<uint8_t> const& bytecode, std::ve
 
     auto composer = AvmComposer();
     auto prover = composer.create_prover(circuit_builder);
-    return prover.construct_proof();
+    auto verifier = composer.create_verifier(circuit_builder);
+    auto proof = prover.construct_proof();
+    return proof;
+}
+
+std::tuple<AvmFlavor::VerificationKey, HonkProof> Execution::prove_vk(std::vector<uint8_t> const& bytecode,
+                                                                      std::vector<FF> const& calldata)
+{
+    auto instructions = Deserialization::parse(bytecode);
+    auto trace = gen_trace(instructions, calldata);
+    auto circuit_builder = bb::AvmCircuitBuilder();
+    circuit_builder.set_trace(std::move(trace));
+
+    auto composer = AvmComposer();
+    auto prover = composer.create_prover(circuit_builder);
+    auto verifier = composer.create_verifier(circuit_builder);
+    auto proof = prover.construct_proof();
+    // todo: Might need to return PCS vk when full verify is supported
+    return std::make_tuple(*verifier.key, proof);
+}
+bool Execution::verify(AvmFlavor::VerificationKey vk, const HonkProof& proof)
+{
+    auto verification_key = std::make_shared<AvmFlavor::VerificationKey>(vk);
+    AvmVerifier verifier(verification_key);
```

Comment:
AvmVerifier takes a shared_ptr... ugh...

Especially in this case you see that the verif key lifetime is as long. Either taking a const ref or a unique_ptr would've made more sense to me.

Related to our talk yesterday: https://stackoverflow.com/questions/7657718/when-to-use-shared-ptr-and-when-to-use-raw-pointers

(btw, nothing for you to change here)